File: python/shared/paths.py
"""Cross-platform path helpers for Voice Mirror Python backend."""

import os
import platform
from pathlib import Path


def get_config_base() -> Path:
    """Get the platform-appropriate base config directory."""
    system = platform.system()
    if system == "Windows":
        base = os.environ.get("APPDATA", str(Path.home() / "AppData" / "Roaming"))
        return Path(base)
    elif system == "Darwin":
        return Path.home() / "Library" / "Application Support"
    else:
        return Path(os.environ.get("XDG_CONFIG_HOME", str(Path.home() / ".config")))


def get_data_dir() -> Path:
    """Get the Voice Mirror data directory (cross-platform)."""
    return get_config_base() / "voice-mirror-electron" / "data"


def safe_path(path, allowed_base) -> str:
    """Resolve *path* and verify it is within *allowed_base*.

    Both arguments are resolved with ``os.path.realpath`` so symlinks
    and ``..`` components are collapsed before the check.

    Returns the resolved path string on success, raises ``ValueError``
    if the path escapes the allowed base directory.
    """
    resolved = os.path.realpath(str(path))
    base = os.path.realpath(str(allowed_base))
    # Allow the base itself or anything under base + separator
    if resolved == base or resolved.startswith(base + os.sep):
        return resolved
    raise ValueError(f"Path {path} is outside allowed directory {allowed_base}")

File: voice-core/src/config/paths.rs
//! Platform-specific data directory paths.
//!
//! Mirrors python/shared/paths.py:
//!   Windows: %APPDATA%/voice-mirror-electron/data
//!   macOS:   ~/Library/Application Support/voice-mirror-electron/data
//!   Linux:   $XDG_CONFIG_HOME/voice-mirror-electron/data (default ~/.config)

use std::path::PathBuf;

/// Get the Voice Mirror data directory (cross-platform).
pub fn get_data_dir() -> PathBuf {
    get_config_base()
        .join("voice-mirror-electron")
        .join("data")
}

/// Get the platform-appropriate base config directory.
fn get_config_base() -> PathBuf {
    #[cfg(target_os = "windows")]
    {
        // %APPDATA% (typically C:\Users\<user>\AppData\Roaming)
        if let Some(appdata) = std::env::var_os("APPDATA") {
            return PathBuf::from(appdata);
        }
        dirs::config_dir().unwrap_or_else(|| {
            dirs::home_dir()
                .unwrap_or_else(|| PathBuf::from("."))
                .join("AppData")
                .join("Roaming")
        })
    }

    #[cfg(target_os = "macos")]
    {
        dirs::home_dir()
            .unwrap_or_else(|| PathBuf::from("."))
            .join("Library")
            .join("Application Support")
    }

    #[cfg(not(any(target_os = "windows", target_os = "macos")))]
    {
        // Linux and other Unix: respect XDG_CONFIG_HOME, default ~/.config
        if let Some(xdg) = std::env::var_os("XDG_CONFIG_HOME") {
            return PathBuf::from(xdg);
        }
        dirs::home_dir()
            .unwrap_or_else(|| PathBuf::from("."))
            .join(".config")
    }
}
